--- a/net/minecraft/world/entity/vehicle/MinecartHopper.java
+++ b/net/minecraft/world/entity/vehicle/MinecartHopper.java
@@ -68,6 +68,13 @@
         this.enabled = enabled;
     }
 
+    // Paper start - add back getLevel
+    @Override
+    public net.minecraft.world.level.Level getLevel() {
+        return this.level;
+    }
+    // Paper end
+
     @Override
     public double getLevelX() {
         return this.getX();
