--- a/net/minecraft/world/entity/Marker.java
+++ b/net/minecraft/world/entity/Marker.java
@@ -49,8 +49,8 @@
     }
 
     @Override
-    protected void addPassenger(Entity passenger) {
-        throw new IllegalStateException("Should never addPassenger without checking couldAcceptPassenger()");
+    protected boolean addPassenger(Entity passenger) { // Paper - fix return type
+        return false; // Paper
     }
 
     @Override
